Update to reflect FrozenTime (#7038)
cakephp/docs: en/core-libraries/time.rst

.. start-time

Creating FrozenTime Instances

> =======================

Reviewer: Title underlines need to be the same length as the text above them.

    use Cake\I18n\FrozenTime;

    // Create from a string datetime.
    $time = FrozenTime::createFromFormat(
        'Y-m-d H:i:s',
        '2021-01-31 22:11:30',
        'America/New_York'
    );

    // Create from a timestamp and set timezone
    $time = FrozenTime::createFromTimestamp(1612149090, 'America/New_York');

    // Get the current time.
    $time = FrozenTime::now();

    // Or just use 'new'
    $time = new FrozenTime('2021-01-31 22:11:30', 'America/New_York');

    $time = new FrozenTime('2 hours ago');

The ``FrozenTime`` class constructor can take any parameter that the internal ``DateTimeImmutable``
PHP class can. When passing a number or numeric string, it will be interpreted
as a UNIX timestamp.

In test cases, you can mock out ``now()`` using ``setTestNow()``::

    // Fixate time.
    $time = new FrozenTime('2021-01-31 22:11:30');
    FrozenTime::setTestNow($time);

    // Outputs '2021-01-31 22:11:30'
    $now = FrozenTime::now();
    echo $now->i18nFormat('yyyy-MM-dd HH:mm:ss');

    // Outputs '2021-01-31 22:11:30'
    $now = FrozenTime::parse('now');
    echo $now->i18nFormat('yyyy-MM-dd HH:mm:ss');

Manipulation
============

> Once created, ``FrozenTime`` instances cannot be manipulated or changed anymore, as they are immutable. Manipulating them will create new instance::

Reviewer 2: This sounds like Schrödinger's time objects again, as if you can, and can not manipulate them.

Author: How about this  :

Once created, ``FrozenTime`` instances cannot be manipulated or changed, as it is immutable. A new instance can be created from a ``FrozenTime`` instance.

.. start-time

Creating FrozenTime Instances
=======================

``FrozenTime`` are immutable objects that useful when you want to prevent
accidental changes to data, or when you want to avoid order based dependency

> issues. Refer to ``Time`` instances for mutable objects. There are a few ways to create ``FrozenTime`` instances::

Reviewer 2: ```suggestion
``FrozenTime`` are immutable objects that are useful when you want to prevent
accidental changes to data, or when you want to avoid order based dependency
issues. Refer to ``Time`` instances for mutable objects.

There are a few ways to create ``FrozenTime`` instances::
```

    use Cake\I18n\FrozenTime;

    // Create from a string datetime.
    $time = FrozenTime::createFromFormat(
        'Y-m-d H:i:s',
        '2021-01-31 22:11:30',
        'America/New_York'
    );

    // Create from a timestamp and set timezone
    $time = FrozenTime::createFromTimestamp(1612149090, 'America/New_York');

    // Get the current time.
    $time = FrozenTime::now();

    // Or just use 'new'
    $time = new FrozenTime('2021-01-31 22:11:30', 'America/New_York');

    $time = new FrozenTime('2 hours ago');

The ``FrozenTime`` class constructor can take any parameter that the internal ``DateTimeImmutable``
PHP class can. When passing a number or numeric string, it will be interpreted
as a UNIX timestamp.

In test cases, you can mock out ``now()`` using ``setTestNow()``::

    // Fixate time.
    $time = new FrozenTime('2021-01-31 22:11:30');
    FrozenTime::setTestNow($time);

    // Outputs '2021-01-31 22:11:30'
    $now = FrozenTime::now();
	echo $now->i18nFormat('yyyy-MM-dd HH:mm:ss');

    // Outputs '2021-01-31 22:11:30'
    $now = FrozenTime::parse('now');

> 	echo $now->i18nFormat('yyyy-MM-dd HH:mm:ss');

Reviewer 2: ```suggestion
    echo $now->i18nFormat('yyyy-MM-dd HH:mm:ss');
```

Manipulation
============

> Once created, ``FrozenTime`` instances cannot be manipulated as it is immutable. You will need to create another instance. You can also use the methods provided by PHP's built-in ``DateTime`` class::

Reviewer 2: ```suggestion
Once created, ``FrozenTime`` instances will not change anymore, as they are immutable. Manipulating them will create new instances that need to be reassigned::
```

>     $time = FrozenTime::parse('2021-01-31 22:11:30');
>     
>     // Create another instance
>     $newTime = $time->setDate(2013, 10, 31);

Reviewer 2: ```suggestion
    $time = FrozenTime::now();
    
    // Create and reassign a new instance
    $time = $now->year(2013)
        ->month(10)
        ->day(31);
```

    // Outputs '2013-10-31 22:11:30'
    echo $newTime->i18nFormat('yyyy-MM-dd HH:mm:ss');

> Using setter methods to manipulate ``FrozenTime`` instances will not work::

Reviewer 2: ```suggestion
You can also use the methods provided by PHP's built-in ``DateTime`` class::

    $time = $time->setDate(2013, 10, 31);

Failing to reassign the new ``FrozenTime`` instances will result in the original, unmodified instance being used::
```

    $time->year(2013)
        ->month(10)
        ->day(31);
    // Outputs '2021-01-31 22:11:30'
    echo $time->i18nFormat('yyyy-MM-dd HH:mm:ss');

You can create another instance with modified dates, through subtraction and addition of their components::

    $time = FrozenTime::create(2021, 1, 31, 22, 11, 30);
    $newTime = $time->subDays(5)

> 		->addHours(-2)

Reviewer 2: ```suggestion
        ->addHours(-2)
```

        ->addMonth(1);
    // Outputs '2/26/21, 8:11 PM'
    echo $newTime;

    // Using strtotime strings.
	$newTime = $time->modify('+1 month -5 days -2 hours');

> 	// Outputs '2/26/21, 8:11 PM'

Reviewer 2: ```suggestion
    $newTime = $time->modify('+1 month -5 days -2 hours');
    // Outputs '2/26/21, 8:11 PM'
```

In test cases you can mock out ``now()`` using ``setTestNow()``::

    // Fixate time.
    $time = new FrozenTime('2021-01-31 22:11:30');
    FrozenTime::setTestNow($time);

    // Outputs '1/31/21, 10:11 PM'
    echo FrozenTime::now();

    // Outputs '1/31/21, 10:11 PM'

>     echo FrozenTime::parse('now');

Reviewer 3: These should output a specific format to match the one used in setTestNow.

Manipulation
============

> Once created, you cannot manipulate ``FrozenTime`` instances as it is immutable. Using setter methods will not work::

Reviewer 4: Paragraph shouldn't start with _you cannot_.

Start with correct example which show you need to assign new instance to variable.

You can insert incorrect example later in paragraph.

Reviewer 3: This is misleading. It sounds like you can't call the methods.

This method sets the default format used when converting an object to json::

>     

Reviewer 2: ```suggestion
```

    Time::setJsonEncodeFormat('yyyy-MM-dd HH:mm:ss');  // For any mutable DateTime
    FrozenTime::setJsonEncodeFormat('yyyy-MM-dd HH:mm:ss');  // For any immutable DateTime
    Date::setJsonEncodeFormat('yyyy-MM-dd HH:mm:ss');  // For any mutable Date
    FrozenDate::setJsonEncodeFormat('yyyy-MM-dd HH:mm:ss');  // For any immutable Date

    $time = FrozenTime::parse('2021-01-31 22:11:30');
    echo json_encode($time);   // Outputs '2021-01-31 22:11:30'
    
    // Added in 4.1.0
    FrozenDate::setJsonEncodeFormat(static function($time) {
        return $time->format(DATE_ATOM);
    });

>     

Reviewer 2: ```suggestion
```

A very common thing to do with ``Time`` instances is to print out formatted
dates. CakePHP makes this a snap::

> 	$time = FrozenTime::parse('2021-01-31 22:11:30');

Reviewer 2: ```suggestion
    $time = FrozenTime::parse('2021-01-31 22:11:30');
```

	// Prints a localized datetime stamp. Outputs '1/31/21, 10:11 PM'

> 	echo $time;

Reviewer 2: ```suggestion
    // Prints a localized datetime stamp. Outputs '1/31/21, 10:11 PM'
    echo $time;
```

    // Outputs '1/31/21, 10:11 PM' for the en-US locale

> 	echo $time->i18nFormat();

Reviewer 2: ```suggestion
    echo $time->i18nFormat();
```

    // Use the full date and time format. Outputs 'Sunday, January 31, 2021 at 10:11:30 PM Eastern Standard Time'

> 	echo $time->i18nFormat(\IntlDateFormatter::FULL);

Reviewer 2: ```suggestion
    echo $time->i18nFormat(\IntlDateFormatter::FULL);
```

You can also format dates with non-gregorian calendars::

    // On ICU version 66.1
    $time = FrozenTime::create(2021, 1, 31, 22, 11, 30);
    
    // Outputs 'Sunday, Bahman 12, 1399 AP at 10:11:30 PM Eastern Standard Time'
    echo $time->i18nFormat(\IntlDateFormatter::FULL, null, 'en-IR@calendar=persian');
    
    // Outputs 'Sunday, January 31, 3 Reiwa at 10:11:30 PM Eastern Standard Time'
    echo $time->i18nFormat(\IntlDateFormatter::FULL, null, 'en-JP@calendar=japanese');
    
    // Outputs 'Sunday, Twelfth Month 19, 2020(geng-zi) at 10:11:30 PM Eastern Standard Time'
    echo $time->i18nFormat(\IntlDateFormatter::FULL, null, 'en-CN@calendar=chinese');
    
    // Outputs 'Sunday, Jumada II 18, 1442 AH at 10:11:30 PM Eastern Standard Time'
    echo $time->i18nFormat(\IntlDateFormatter::FULL, null, 'en-SA@calendar=islamic');
    

>     

Reviewer 2: ```suggestion
```

    FrozenTime::setDefaultLocale('es-ES'); // For any immutable DateTime
    Date::setDefaultLocale('es-ES'); // For any mutable Date
    FrozenDate::setDefaultLocale('es-ES'); // For any immutable Date
    
    // Outputs '31 ene. 2021 22:11'
    echo $time->nice(); 

>     

Reviewer 2: ```suggestion
```

    Date::setToStringFormat(\IntlDateFormatter::SHORT); // For any mutable Date
    FrozenDate::setToStringFormat(\IntlDateFormatter::SHORT); // For any immutable Date

>     // The same method exists on Date, FrozenDate, Time and FrozenTime

Reviewer 2: ```suggestion
    // The same method exists on Date, FrozenDate, and Time
```

    FrozenTime::setToStringFormat([
        \IntlDateFormatter::FULL,
        \IntlDateFormatter::SHORT
    ]);
    // Outputs 'Sunday, January 31, 2021 at 10:11 PM'
    echo $time; 

>     // The same method exists on Date, FrozenDate, Time and FrozenTime

Reviewer 2: ```suggestion
    // The same method exists on Date, FrozenDate, and Time
```

Often it is useful to print times relative to the present::

    $time = new FrozenTime('Jan 31, 2021');

>     // On 12-Jun-21, outputs '4 months, 1 week, 6 days ago'

Reviewer 2: ```suggestion
    // On June 12, 2021, this would output '4 months, 1 week, 6 days ago'
```

.. php:method:: isThisMonth()
.. php:method:: isThisYear()

You can compare a ``FrozenTime`` instance with the present in a variety of ways::

    $time = new FrozenTime('+3 days');

    var_export($time->isYesterday());
    var_export($time->isThisWeek());
    var_export($time->isThisMonth());
    var_export($time->isThisYear());

Each of the above methods will return ``true``/``false`` based on whether or
not the ``FrozenTime`` instance matches the present.

Comparing With Intervals
========================

.. php:method:: isWithinNext($interval)

You can see if a ``FrozenTime`` instance falls within a given range using
``wasWithinLast()`` and ``isWithinNext()``::

    $time = new FrozenTime('+3 days');

    // Within 2 days. Outputs 'false'
    var_export($time->isWithinNext('2 days'));

    // Within 2 next weeks. Outputs 'true'

>      var_export($time->isWithinNext('2 weeks'));

Reviewer 2: ```suggestion
    var_export($time->isWithinNext('2 weeks'));
```

You can also compare a ``FrozenTime`` instance within a range in the past::

    $time = new FrozenTime('-72 hours');
    
    // Within past 2 days. Outputs 'false'
    var_export($time->wasWithinLast('2 days'));

    // Within past 3 days. Outputs 'true'
    var_export($time->wasWithinLast('3 days'));
    
    // Within past 2 weeks. Outputs 'true'
    var_export($time->wasWithinLast('2 weeks'));

.. end-time

FrozenDate
=====

.. php:class: FrozenDate

The immutable ``FrozenDate`` class in CakePHP implements the same API and methods as
:php:class:`Cake\\I18n\\FrozenTime` does. The main difference between ``FrozenTime`` and
``FrozenDate`` is that ``FrozenDate`` does not track time components, and is always in UTC.
As an example::

    use Cake\I18n\FrozenDate;
    $date = new FrozenDate('2021-01-31');

    $newDate = $date->modify('+2 hours');
    // Outputs '2021-01-31 00:00:00'
    echo $newDate->format('Y-m-d H:i:s');
    
    $newDate = $date->addHours(36);
    // Outputs '2021-01-31 00:00:00'
    echo $newDate->format('Y-m-d H:i:s');

    $newDate = $date->addDays(10);
    // Outputs '2021-02-10 00:00:00'
    echo $newDate->format('Y-m-d H:i:s');

>     

Reviewer 2: ```suggestion
```

.. php:method:: isThisMonth()
.. php:method:: isThisYear()

You can compare a ``FrozenTime`` instance with the present in a variety of ways::

    $time = new FrozenTime('+3 days');

>     var_export($time->isYesterday());
>     var_export($time->isThisWeek());
>     var_export($time->isThisMonth());
>     var_export($time->isThisYear());

Reviewer: ```suggestion
    debug($time->isYesterday());
    debug($time->isThisWeek());
    debug($time->isThisMonth());
    debug($time->isThisYear());
```

I don't think `var_export` is what we want here as we're not trying to create executable code. `debug()` will also print the true/false properly and is more inline with the rest of the cake docs.

Each of the above methods will return ``true``/``false`` based on whether or
not the ``FrozenTime`` instance matches the present.

Comparing With Intervals
========================

.. php:method:: isWithinNext($interval)

You can see if a ``FrozenTime`` instance falls within a given range using
``wasWithinLast()`` and ``isWithinNext()``::

    $time = new FrozenTime('+3 days');

    // Within 2 days. Outputs 'false'

>     var_export($time->isWithinNext('2 days'));

Reviewer: ```suggestion
    debug($time->isWithinNext('2 days'));
```

    // Within 2 next weeks. Outputs 'true'

>     var_export($time->isWithinNext('2 weeks'));

Reviewer: ```suggestion
    debug($time->isWithinNext('2 weeks'));
```

.. php:method:: wasWithinLast($interval)

You can also compare a ``FrozenTime`` instance within a range in the past::

    $time = new FrozenTime('-72 hours');
    
    // Within past 2 days. Outputs 'false'

>     var_export($time->wasWithinLast('2 days'));

Reviewer: ```suggestion
    debug($time->wasWithinLast('2 days'));
```

    // Within past 3 days. Outputs 'true'

>     var_export($time->wasWithinLast('3 days'));

Reviewer: ```suggestion
    debug($time->wasWithinLast('3 days'));
```

    
    // Within past 2 weeks. Outputs 'true'

>     var_export($time->wasWithinLast('2 weeks'));

Reviewer: ```suggestion
    debug($time->wasWithinLast('2 weeks'));
```

.. end-time

FrozenDate

> =====

Reviewer: Title underlines need to be the same length as the text above them.

.. php:class: FrozenDate

The immutable ``FrozenDate`` class in CakePHP implements the same API and methods as
:php:class:`Cake\\I18n\\FrozenTime` does. The main difference between ``FrozenTime`` and

> ``FrozenDate`` is that ``FrozenDate`` does not track time components, and is always in UTC.

Reviewer 3: ```suggestion
``FrozenDate`` is that ``FrozenDate`` does not track time components.
```
This is no longer true, but it's an implementation detail that shouldn't matter.

Author: ok, will remove it from next PR. Can't change it now cause already merged.
